sematic/plugins/storage/tests/test_local_storage.py

```diff
@@ -23,9 +23,8 @@ def test_upload_download(
         with mock.patch(
             "sematic.plugins.storage.local_storage._get_data_dir", return_value=tempdir
         ):
-            response = test_client.put(
-                local_storage.get_write_location("artifacts", "123"), data=value
-            )
+            location = local_storage.get_write_location("artifacts", "123")
+            response = test_client.put(f"/api/v1{location}", data=value)
```

Comment:
Please correct me if I understand it wrong.

This way, the client might have to prepend `/api/v1` by itself when writing to storage. How could the client have this knowledge, without knowing the type of storage being used? If the storage is S3 storage, we likely don't have to prepend `/api/v1` right? If so, this design might not be storage-type agnostic.

Reply:
See my comment in `api_client`.
In the case of API calls, the client needs to add authentication headers, so we can't blindly query whatever is returned.
The contract is that external URLs (e.g. S3) are returned as fully qualified URL, whereas Sematic API endpoints are returned as relative URLs (e.g. `/artifact/<id>/data`). Then the client knows what to do.

A better long-term solution would be for the server to return both the URL and the necessary headers. In that case, we would not need the logic in `api_client`. I will make the change in a follow-up PR.

Reply:
Thank you for sharing this thinking, @neutralino1. I have the same feeling that there might be room to simplify. Also, I guess it might be helpful to consider the following two alternatives:

1. Return a signed URL as S3 does. The signed URL encodes a secret which consists of a. the semantic API_Key, b. insurance timestamp. When servicing the signed URL, the server unpacks the secret and checks API_Key and timestamp. Then this endpoint needs to bypass the ordinary API_Key checking mechanism but look at the secret in the URL instead. Also, the server only serves if the timestamp has expired for a short time. This avoids some complexities of manipulating headers for the client. (Honestly, I haven't checked how S3 does the signed URL, if there is a better way there, we can also reference)

2. We can also use HTTP interceptors in the client; The interceptor is responsible for attaching credentials to HTTP request headers based on different service endpoints. So we decouple the issues at different places, also relieving the usage burden of the client.